### DISCUSSION TITLE
Arroost capabilities

### DISCUSSION THREAD

**Discussion by Jeremy5909** [TodePond/Arroost#156] [General]
> Arroost is a ridiculous music-making tool.

Is this true or just a part of it? I recall you saying somewhere earlier that this will be capable of doing everything CellPond can and more, does this still hold up? Will it also be able to make that sand splitting thing? I'll find the name of what I'm talking about.

**Comment by loglot** [comment]
Sand lab maybe is what you are thinking about?

**Reply by Jeremy5909** [reply]
Yes - thank you

**Comment by TodePond** [comment]
Hey there's a couple things going on here:

- The underlying VM of Arroost (Nogan) could be used to implement CellPond's VM (Dragon), with some bonus features like time travel. This is because Nogan is a node-based network, which can be used to implement grid-like things (like CellPond).
- The UI engine for Arroost is basically a work-in-progress 'sequel' to the UI engine I made for CellPond. It's like a mini game engine. It still has a bunch of quirks and issues but it's getting better. I want to be able to make this kind of 'visual coding UI' very quickly, and Arroost is part of my continual work towards that 
- SandLab is an experimental 'variation' of CellPond. In it, I try to allow the user to make rules that are more flexible. In CellPond, the main primitive of a rule is a "cell". And they all have to be 'lined up' and in proportion to each other. In SandLab, the main primitive of a rule is an "edge". You define rules in terms of how edges and contact points should interact. You can control how strict or loose the pattern matching has to be.

I've done five main stints of work on CellPond so it's very mature at this point. Arroost will be at a very early stage still, in the first video. These things take a long time! But from seeing people's positive reactions to CellPond at this later stage, I feel very reassured that it's worth it! Even if it's less interesting at the beginning

**Reply by loglot** [reply]
So what you're saying is, there is a chance that we will be able to make a computer

**Reply by Jeremy5909** [reply]
So to confirm will everything in cell pond be possible in arroost? or just in Nogan. Very excited

**Reply by TodePond** [reply]
No, it won't. It would be an oversimplification to say that.

- Nogan is a library. You could build Dragon in it (CellPond's VM) but I won't do that for a long long time.
- Arroost uses a user interface library. You could use that user interface library to rebuild Cellpond's user interface. But I won't do that for a long long time.

**Reply by mahdoosh1** [reply]
> So what you're saying is, there is a chance that we will be able to make a computer 

I have suggested that and you are aware. But any type of negative logic can allow us to make computers. Currently we have `or` gate as default behavior of a node.

**Reply by Jeremy5909** [reply]
still a bit confused on how you can make Dragon in Nogan 😭  but I'm figuring it out lol

**Reply by TodePond** [reply]
> still a bit confused on how you can make Dragon in Nogan 😭 but I'm figuring it out lol

essentially, both are low level specifications of rewrite rules. nogan is lower level than dragon so you could specify dragon in terms of nogan. the one alteration you need to make is that nogan is highly deterministic whereas dragon is highly random. that's not required though - it's implementation specific

**Reply by Jeremy5909** [reply]
Alright I think I understand, thank you!
